#13127: Clean up layout conversion

TT-BrianLiu opened the pull request:
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/13127)

### Problem description
Cleaning up how we handle data conversion for python <-> tt tensors.

### What's changed
- Deprecate support for ROW_MAJOR BFLOAT8_B/BFLOAT4_B tensors on host
  - Throw warning when users try to convert layout for BFLOAT8_B/BFLOAT4_B dtypes
  - Throw warning and return original input tensor instead of assert in pad op if input is not ROW_MAJOR layout
    - This is a workaround to support old test code that tries to:
       1. Create BFLOAT8_B tensors in ROW_MAJOR
       2. Pad up to tile sizes
       3. Convert layout to TILE
    - Now, we just directly pad and return in TILE layout on the first step
  - Update tensor creation with float vectors to always output TILE layout for BFLOAT8_B/BFLOAT4_B dtypes
    - Automatically add padding to tile sizes if needed
    - Throw warning if requested layout is not TILE
    - Remove duplicate code by calling a helper function for the different pybinds
    - Remove support for BFLOAT8_B/BFLOAT4_B in create_owned_buffer_from_vector_of_floats
  - Update conv weights utils to do layout conversion in FLOAT32, then unpack and re-pack into BFLOAT8_B/BFLOAT4_B
    - This is strictly better for host-side perf since we eliminate one unecessary pack
  - Remove duplicate code in handling of BFLOAT8_B/BFLOAT4_B dtypes in pytensor
- Update convert_layout to use 2D physical shape for all underlying conversions
  - Top level convert layout has a pass to flatten shape for existing use cases
  - Remove padding support for tilize_nchw since this is unused at this level
  - Remove recursive calls inside convert layout to simplify logic
  - Switch convert_layout_row_major_to_tile and convert_layout_tile_to_row_major to use 2D physical shape
  - Update tests/tt_metal/tt_metal/test_bcast.cpp to handle padding for tile externally

### Checklist
- [x] Post commit CI passes
  - Post-commit all: https://github.com/tenstorrent/tt-metal/actions/runs/12110164523
  - Nightly fast dispatch: https://github.com/tenstorrent/tt-metal/actions/runs/12037520925
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

ayerofieiev-tt commented on tt_metal/common/test_tiles.hpp:
```diff
@@ -353,3 +334,23 @@ inline std::vector<T> convert_layout(
+    ZoneScoped;
+
+    TT_ASSERT(shape.size() >= 2, "Shape size {} must be at least rank 2!", shape.size());
+    uint32_t H = shape[shape.size() - 2], W = shape[shape.size() - 1];
```
are you sure uint32_t here is enough?

TT-BrianLiu replied:
absolutely. We don't have enough memory to store a tensor that large

ayerofieiev-tt commented on tt_metal/common/test_tiles.hpp:
```diff
@@ -353,3 +334,23 @@ inline std::vector<T> convert_layout(
+    ZoneScoped;
+
+    TT_ASSERT(shape.size() >= 2, "Shape size {} must be at least rank 2!", shape.size());
+    uint32_t H = shape[shape.size() - 2], W = shape[shape.size() - 1];
```
please split into 2 lines

ayerofieiev-tt commented on ttnn/cpp/pybind11/pytensor.cpp:
```diff
@@ -76,6 +76,59 @@ Tensor create_owned_tensor(
+            return owned_buffer::create<float>(std::move(data));
+        }
+        case DataType::BFLOAT16: {
+            std::vector<::bfloat16> bfloat16_data(data.size());
+            std::transform(std::begin(data), std::end(data), std::begin(bfloat16_data), [](float value) {
+                return ::bfloat16(value);
+            });
+            return owned_buffer::create<::bfloat16>(std::move(bfloat16_data));
```
Does this conversion have to be value by value? Just curious

TT-BrianLiu replied:
not sure. I moved existing code

ayerofieiev-tt commented on ttnn/cpp/pybind11/pytensor.cpp:
```diff
@@ -76,6 +76,59 @@ Tensor create_owned_tensor(
+            });
+            return owned_buffer::create<::bfloat16>(std::move(bfloat16_data));
+        }
+        default: {
+            TT_THROW("Cannot create a host buffer!");
+        }
```
what about integer types?

TT-BrianLiu replied:
Old legacy code path that I am not sure we will need.

ayerofieiev-tt commented on ttnn/cpp/pybind11/pytensor.cpp:
```diff
@@ -76,6 +76,59 @@ Tensor create_owned_tensor(
     return Tensor(std::move(storage), shape, data_type, layout, optional_tile);
 }
 
+OwnedBuffer create_owned_buffer_from_vector_of_floats(std::vector<float> &&data, DataType data_type) {
```
why `std::vector<float> &&data`?

TT-BrianLiu replied:
const ref better?

ayerofieiev-tt commented on ttnn/cpp/pybind11/pytensor.cpp:
```diff
@@ -76,6 +76,59 @@ Tensor create_owned_tensor(
+            log_warning(tt::LogAlways, "Tensor layout must be Layout::TILE for bfloat8_b or bfloat4_b! Tensor layout will be {} instead of the requested {}!", Layout::TILE, layout);
+        }
+        auto owned_buffer = create_owned_buffer_from_vector_of_floats(std::move(data), DataType::FLOAT32);
+        auto float_tensor = Tensor(OwnedStorage{owned_buffer}, shape, DataType::FLOAT32, Layout::ROW_MAJOR, tile).to(Layout::TILE);
```
can we avoid creating ROW_MAJOR tensor with `tile` argument?

TT-BrianLiu replied:
Out of scope for this PR. We need make `to_layout` take in tile arg if it's ROW_MAJOR.

ayerofieiev-tt commented on tt_metal/common/test_tiles.hpp:
```diff
@@ -353,3 +334,24 @@ inline std::vector<T> convert_layout(
+template <typename T, template <typename...> typename BufferType>
+inline std::vector<T> convert_layout(
+    const BufferType<T>& inp,
+    tt::stl::Span<const uint32_t> shape,
```
should we pass SimpleShape here?

TT-BrianLiu replied:
I don't know if we want to introduce dependencies on ttnn in tt_metal code

TT-BrianLiu committed:
#13127: Clean up layout conversion
- Deprecate support for ROW_MAJOR BFLOAT8_B/BFLOAT4_B tensors on host
  * Throw warning when users try to convert layout for BFLOAT8_B/BFLOAT4_B dtypes
  * Throw warning and return original input tensor instead of assert in pad op if input is not ROW_MAJOR layout
    ** This is a workaround to support old test code that tries to:
       1. Create BFLOAT8_B tensors in ROW_MAJOR
       2. Pad up to tile sizes
       3. Convert layout to TILE
    ** Now, we just directly pad and return in TILE layout on the first step
  * Update tensor creation with float vectors to always output TILE layout for BFLOAT8_B/BFLOAT4_B dtypes
    ** Automatically add padding to tile sizes if needed
    ** Throw warning if requested layout is not TILE
    ** Remove duplicate code by calling a helper function for the different pybinds
    ** Remove support for BFLOAT8_B/BFLOAT4_B in create_owned_buffer_from_vector_of_floats
  * Update conv weights utils to do layout conversion in FLOAT32, then unpack and re-pack into BFLOAT8_B/BFLOAT4_B
    ** This is strictly better for host-side perf since we eliminate one unecessary pack
  * Remove duplicate code in handling of BFLOAT8_B/BFLOAT4_B dtypes in pytensor
- Update convert_layout to use 2D physical shape for all underlying conversions
  * Top level convert layout has a pass to flatten shape for existing use cases
  * Remove padding support for tilize_nchw since this is unused at this level
  * Remove recursive calls inside convert layout to simplify logic
  * Switch convert_layout_row_major_to_tile and convert_layout_tile_to_row_major to use 2D physical shape
  * Update tests/tt_metal/tt_metal/test_bcast.cpp to handle padding for tile externally
- Fix segfault when calling tensor creation with float vectors with device = None
